src/lib/optimizer/strategy/like_replacement_rule.cpp

```diff
@@ -0,0 +1,73 @@
+
+std::string LikeReplacementRule::name() const { return "Like Replacement Rule"; }
+
+void LikeReplacementRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
+  if (predicate_node != nullptr) {
+    const auto expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate_node->predicate());
+    if (expression != nullptr && expression->predicate_condition == PredicateCondition::Like) {
+      const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->left_operand());
+      const auto value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->right_operand());
+      if (value_expression != nullptr && column_expression != nullptr) {
+        const auto value = boost::get<std::string>(value_expression->value);
+        const auto offset = value.find_first_of("%");
+        // Check if the string has an % wildcard and ends with the % wildcard
+        if (offset != std::string::npos && offset > 0 && value.length() == offset + 1) {
+          // Calculate lower and upper bound of the string
+          const auto lower_bound = value.substr(0, offset);
+          const auto current_character_value = static_cast<int>(lower_bound.at(lower_bound.length() - 1));
+          // Find next value according to ASCII-table
+          if (current_character_value > 0 && current_character_value < 127) {
+            const auto next_character = static_cast<char>(current_character_value + 1);
+            const auto upper_bound = lower_bound.substr(0, offset - 1) + next_character;
+            const auto lower_bound_node = PredicateNode::make(
+                std::make_shared<BinaryPredicateExpression>(PredicateCondition::GreaterThanEquals, column_expression,
+                                                            std::make_shared<ValueExpression>(lower_bound)));
+            const auto upper_bound_node = PredicateNode::make(std::make_shared<BinaryPredicateExpression>(
+                PredicateCondition::LessThan, column_expression, std::make_shared<ValueExpression>(upper_bound)));
+
+            // Replace like node with boundary nodes
+
+            // Store the input and outputs of the node
+            auto input = predicate_node->left_input();
+            const auto outputs = predicate_node->outputs();
+            const auto input_sides = predicate_node->get_input_sides();
+
+            lqp_remove_node(predicate_node);
+
+            // Connect the boundary nodes with the input and outputs of the replaced like node
+            lower_bound_node->set_left_input(input);
+            upper_bound_node->set_left_input(lower_bound_node);
+            for (size_t output_idx = 0; output_idx < outputs.size(); ++output_idx) {
+              outputs[output_idx]->set_input(input_sides[output_idx], upper_bound_node);
+            }
+          }
+        }
+      }
+    }
+  }
+  _apply_to_inputs(node);
```

Comment:
I think you might have an unintended abort condition here.
From my understanding, you want this rule to be applied to all `LIKE` nodes. However, I think it is applied only to the first `LIKE` node in the LQP.

Reasoning:
L. 26 declares `predicate_node` as type-casted version of  `node`, so they should both point to the same object.
L. 57 calls `lqp_remove_node(predicate_node)`, which sets the (left) input of that node to `nullptr`.
L. 70 calls `_apply_to_inputs(node)`, which applies the rule to all inputs. However, since `node` references the same object as `predicate_node`, it has no longer any inputs, and the recursion stops.
I must admit that I did not wrote a test to prove it, however I think it would be good to have a test that covers the double-like-edgecase.

To fix this, one could maybe pull this line up, or save the original input in a variable and apply the rule to it.

Reply:
Thanks a lot!